archive/tar_stream.ts

```diff
@@ -0,0 +1,466 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+
+/**
+ * The interface required to provide a file.
+ */
+export interface TarStreamFile {
+  /**
+   * The name of the file.
+   */
+  pathname: string;
+  /**
+   * The size of the file.
+   */
+  size: number;
+  /**
+   * The contents of the file.
+   */
+  readable: ReadableStream<Uint8Array>;
+  /**
+   * The metadata of the file.
+   */
+  options?: TarStreamOptions;
+}
+
+/**
+ * The interface required to provide a directory.
+ */
+export interface TarStreamDir {
+  /**
+   * The name of the directory.
+   */
+  pathname: string;
+  /**
+   * The metadata of the directory.
+   */
+  options?: TarStreamOptions;
+}
+
+/**
+ * A union type merging all the TarStream interfaces that can be piped into the
+ * TarStream class.
+ */
+export type TarStreamInput = TarStreamFile | TarStreamDir;
+type TarStreamInputInternal =
+  & (Omit<TarStreamFile, "pathname"> | Omit<TarStreamDir, "pathname">)
+  & { pathname: [Uint8Array, Uint8Array] };
+
+/**
+ * The options that can go along with a file or directory.
+ */
+export interface TarStreamOptions {
+  /**
+   * An octal literal.
+   * Defaults to 755 for directories and 644 for files.
+   */
+  mode?: number;
+  /**
+   * An octal literal.
+   * @default {0}
+   */
+  uid?: number;
+  /**
+   * An octal literal.
+   * @default {0}
+   */
+  gid?: number;
```

Comment:
Perhaps we should use number in octal syntax instead (e.g. `0o755` instead of `755`).

Reply:
If people think that's a better experience then I'm open to it. I think it might lead to confusion though:
- with developers not being familiar with octal syntax, and;
- if one console logged `0o755` they'd see `493`, which is less clear what value this means.

Reply:
I saw that the current implementations of Tar and Untar also use octal numbers:

https://github.com/denoland/std/blob/74a92c4acd8e26be8dedd3d63e8a8fee832efda5/archive/tar.ts#L349-L350

https://github.com/denoland/std/blob/74a92c4acd8e26be8dedd3d63e8a8fee832efda5/archive/untar.ts#L436-L442

Reply:
This one needs more discussions, IMO. Just leave it as is for now. 👍

Reply:
Please feel free to open an issue for this for further discussion, if needed.